packages/main/src/plugin/authentication.ts

```diff
@@ -0,0 +1,189 @@
+  id: string;
+  label: string;
+  provider: AuthenticationProvider;
+  options: AuthenticationProviderOptions;
+}
+
+export interface AuthenticationProviderInfo {
+  id: string;
+  displayName: string;
+  accounts: AuthenticationSessionAccountInformation[];
+}
+
+export interface ExtensionInfo {
+  id: string;
+  label: string;
+}
+
+export interface AllowedExtension {
+  id: string;
+  name: string;
+  allowed?: boolean;
+}
+
+export class AuthenticationImpl {
+  private _authenticationProviders: Map<string, ProviderWithMetadata> = new Map<string, ProviderWithMetadata>();
+
+  constructor(private apiSender: any) {}
+
+  public async getAuthenticationProvidersInfo(): Promise<AuthenticationProviderInfo[]> {
+    const values = Array.from(this._authenticationProviders.values());
+    const sessionsRequests = values.map(meta => {
+      return meta.provider.getSessions().then(sessions => {
+        return {
+          id: meta.id,
+          displayName: meta.label,
+          accounts: sessions.map(session => ({ id: session.id, label: session.account.label })),
+        };
+      });
+    });
+
+    return await Promise.all(sessionsRequests);
+  }
+
+  public async signOut(providerId: string, sessionId: string) {
+    this.removeSession(providerId, sessionId);
+  }
+
+  registerAuthenticationProvider(
+    id: string,
+    label: string,
+    provider: AuthenticationProvider,
+    options?: AuthenticationProviderOptions,
+  ): Disposable {
+    if (this._authenticationProviders.get(id)) {
+      throw new Error(`An authentication provider with id '${id}' is already registered.`);
+    }
+    this._authenticationProviders.set(id, {
+      id,
+      label,
+      provider,
+      options: options ?? { supportsMultipleAccounts: false },
+    });
+    this.apiSender.send('authentication-provider-update');
+    const onDidChangeSessionDisposable = provider.onDidChangeSessions(
+      (event: AuthenticationProviderAuthenticationSessionsChangeEvent) => {
+        this._onDidChangeSessions.fire({ provider: { id, label } });
+        this.apiSender.send('authentication-provider-update');
+      },
+    );
+    return {
+      dispose: () => {
+        onDidChangeSessionDisposable.dispose();
+        this._authenticationProviders.delete(id);
+      },
+    };
+  }
+
+  addAccountUsage(providerId: string, accountLabel: string, extensionId: string, extensionName: string): void {
+    throw new Error('The method is not implemented!');
+  }
+
+  readAllowedExtensions(providerId: string, accountName: string): AllowedExtension[] {
+    throw new Error('The method is not implemented!');
+  }
+
+  updateAllowedExtension(
+    providerId: string,
+    accountName: string,
+    extensionId: string,
+    extensionName: string,
+    isAllowed: boolean,
+  ): void {
+    throw new Error('The method is not implemented!');
+  }
+
+  isAccessAllowed(providerId: string, accountName: string, extensionId: string): boolean | undefined {
```

Comment:
in which case it will return undefined ?

Reply:
/**
	 * Check extension access to an account
	 * @param providerId The id of the authentication provider
	 * @param accountName The account name that access is checked for
	 * @param extensionId The id of the extension requesting access
	 * @returns Returns true or false if the user has opted to permanently grant or disallow access, and undefined
	 * if they haven't made a choice yet
	 */

Reply:
I'll add original comments.